aptos-move/framework/src/natives/cryptography/groups/mod.rs

```diff
@@ -0,0 +1,368 @@
+    let g2_handles = pop_arg!(args, Vec<u64>);
+    let g1_handles = pop_arg!(args, Vec<u64>);
+    match (g1_opt, g2_opt, gt_opt) {
+        (Some(Structure::BLS12_381_G1), Some(Structure::BLS12_381_G2), Some(Structure::BLS12_381_Gt)) => {
+            let g1_prepared: Vec<ark_ec::models::bls12::g1::G1Prepared<Parameters>> = g1_handles
+                .iter()
+                .map(|&handle| {
+                    let element = borrow_bls12_381_g1!(context, handle as usize);
+                    ark_ec::prepare_g1::<ark_bls12_381::Bls12_381>(element.into_affine())
+                })
+                .collect();
+            let g2_prepared: Vec<ark_ec::models::bls12::g2::G2Prepared<Parameters>> = g2_handles
+                .iter()
+                .map(|&handle| {
+                    let element = borrow_bls12_381_g2!(context, handle as usize);
+                    ark_ec::prepare_g2::<ark_bls12_381::Bls12_381>(element.into_affine())
+                })
+                .collect();
+
+            let z: Vec<(
+                ark_ec::models::bls12::g1::G1Prepared<Parameters>,
+                ark_ec::models::bls12::g2::G2Prepared<Parameters>,
+            )> = g1_prepared
+                .into_iter()
+                .zip(g2_prepared.into_iter())
+                .collect();
+            let pairing_result = ark_bls12_381::Bls12_381::product_of_pairings(z.as_slice());
+            let result_handle = store_bls12_381_gt!(context, pairing_result);
+            Ok(NativeResult::ok(
+                gas_params.base,
+                smallvec![Value::u64(result_handle as u64)],
+            ))
+        }
+        _ => {
+            Ok(NativeResult::err(InternalGas::zero(), NOT_IMPLEMENTED))
+        }
+    }
+}
+
+pub fn make_all(gas_params: GasParameters) -> impl Iterator<Item = (String, NativeFunction)> {
+    let mut natives = vec![];
+
+    // Always-on natives.
+    natives.append(&mut vec![
+        (
+            "generator_internal",
+            make_native_from_func(gas_params.clone(), generator_internal),
+        ),
+        (
+            "eq_internal",
+            make_native_from_func(gas_params.clone(), eq_internal),
+        ),
+        (
+            "pairing_product_internal",
+            make_native_from_func(gas_params.clone(), pairing_product_internal),
+        ),
+    ]);
+
+    crate::natives::helpers::make_module_natives(natives)
+}
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum Structure {
+    BLS12_381_G1,
+    BLS12_381_G2,
+    BLS12_381_Gt,
+    Ristretto255,
+}
+
+fn structure_from_type_tag(type_tag: &TypeTag) -> Option<Structure> {
+    match type_tag.to_string().as_str() {
+        "0x1::groups::BLS12_381_G1" => Some(Structure::BLS12_381_G1),
+        "0x1::groups::BLS12_381_G2" => Some(Structure::BLS12_381_G2),
+        "0x1::groups::BLS12_381_Gt" => Some(Structure::BLS12_381_Gt),
+        "0x1::groups::Ristretto255" => Some(Structure::Ristretto255),
```

Comment:
For `func<G>` we need to check the actual `G`. Can we rely on this path format, or there's a better way?